### PR TITLE
Migration for direct award random IDs

### PR DESCRIPTION
**Pull request by samuelhwilliams**
Migration PR for https://github.com/alphagov/digitalmarketplace-api/pull/690

### CODE REVIEW

File: migrations/versions/1060_add_external_id_column.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+"""Add external ID column
+
+Revision ID: 1050
```
Comment (TheDoubleK) [inline migrations/versions/1060_add_external_id_column.py]: These should ideally match the ID numbers used by Alembic defined below :trollface:

File: migrations/versions/1060_add_external_id_column.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+DIRECT_AWARD_AUDIT_TYPES = ('create_project', 'create_project_search', 'lock_project', 'downloaded_project')
+
+
+def random_positive_external_id():
+    return random.SystemRandom().randint(10 ** 14, (10 ** 15) - 1)
+
+
+def upgrade():
+    op.add_column('direct_award_projects', sa.Column('external_id', sa.BIGINT()))
```
Comment (TheDoubleK) [inline migrations/versions/1060_add_external_id_column.py]: `supplier_id` was defined as `sa.BigInteger()` - I have no idea if there's any difference between these things, but maybe use that here for consistency?

File: migrations/versions/1060_add_external_id_column.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    op.create_index(op.f('ix_direct_award_projects_external_id'), 'direct_award_projects', ['external_id'], unique=True)
+
+    # Manually generate random IDs for all existing saved searches
+    project_connection = op.get_bind()
+    project_id_internal_to_external = {}
+    for row_id, *_ in project_connection.execute('select id from direct_award_projects'):
+        # Some logic to ensure all IDs generated are unique; keep track so we can update AuditEvents.
+        random_id = random_positive_external_id()
+        while random_id in project_id_internal_to_external:
```
Comment (TheDoubleK) [inline migrations/versions/1060_add_external_id_column.py]: This still won't work with a dict - should be `project_id_internal_to_external.values()`.

**Comment by samuelhwilliams** [comment]
Revised @TheDoubleK

**Comment by galund** [comment]
Do we still have a justification for putting additive migrations into a separate PR?

File: migrations/versions/1070_add_external_id_column.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        new_audit_data = json.dumps({'projectExternalId': project_id_internal_to_external[project_id], **audit_data})
+        audit_connection.execute(text('UPDATE audit_events SET data = (:new_audit_data)::json WHERE id = :audit_id'),
+                                 {'new_audit_data': new_audit_data, 'audit_id': audit_id})
+
+
+def downgrade():
+    op.drop_index(op.f('ix_direct_award_projects_external_id'), table_name='direct_award_projects')
+    op.drop_column('direct_award_projects', 'external_id')
+
+    # I don't think we ought to delete the external ID from the Audit Event on a downgrade... But if we do, here's
```
Comment (TheDoubleK) [inline migrations/versions/1070_add_external_id_column.py]: Let's leave these in on the downgrade.  🙏  we never have to actually do it.

**Comment by samuelhwilliams** [comment]
Closing this as we don't need the separate PR any more. MIgration remains in https://github.com/alphagov/digitalmarketplace-api/pull/690

**Comment by samuelhwilliams** [comment]
Closing this as we don't need the separate PR any more. MIgration remains in https://github.com/alphagov/digitalmarketplace-api/pull/690

**Pull request closed without merging**